File: apps/elixir/test/fizz_buzz/type_test.exs
defmodule FizzBuzz.TypeTest do
  use ExUnit.Case

  alias FizzBuzz.Type
  alias FizzBuzz.Type.{Type01, Type02, Type03, Generatable}

  describe "Type01" do
    test "1 を渡すと 1 を返す" do
      assert Generatable.generate(%Type01{}, 1) == "1"
    end

    test "3 の倍数を渡すと Fizz を返す" do
      assert Generatable.generate(%Type01{}, 3) == "Fizz"
    end

    test "5 の倍数を渡すと Buzz を返す" do
      assert Generatable.generate(%Type01{}, 5) == "Buzz"
    end

    test "15 の倍数を渡すと FizzBuzz を返す" do
      assert Generatable.generate(%Type01{}, 15) == "FizzBuzz"
    end
  end

  describe "Type02" do
    test "数値を文字列に変換する" do
      assert Generatable.generate(%Type02{}, 1) == "1"
    end
  end

  describe "Type03" do
    test "1 を渡すと 1 を返す" do
      assert Generatable.generate(%Type03{}, 1) == "1"
    end

    test "3 を渡すと Fizz を返す" do
      assert Generatable.generate(%Type03{}, 3) == "Fizz"
    end

    test "15 の倍数で FizzBuzz を返す" do
      assert Generatable.generate(%Type03{}, 15) == "FizzBuzz"
    end
  end

  describe "create/1" do
    test "タイプ 1 を生成できる" do
      assert Type.create(1) == {:ok, %Type01{}}
    end

    test "タイプ 2 を生成できる" do
      assert Type.create(2) == {:ok, %Type02{}}
    end

    test "タイプ 3 を生成できる" do
      assert Type.create(3) == {:ok, %Type03{}}
    end

    test "未定義のタイプでエラーを返す" do
      assert Type.create(4) == {:error, "未定義のタイプです"}
    end
  end
end

File: apps/clojure/test/fizzbuzz/domain/type_test.clj
(ns fizzbuzz.domain.type-test
  (:require [clojure.test :refer :all]
            [fizzbuzz.domain.model :as model]
            [fizzbuzz.domain.type :as type]))

(deftest fizz-buzz-type01-test
  (testing "タイプ 1: 通常の FizzBuzz"
    (let [t (type/->FizzBuzzType01)]
      (is (= "1" (type/generate-string t (model/create-fizz-buzz-value 1))))
      (is (= "2" (type/generate-string t (model/create-fizz-buzz-value 2))))
      (is (= "Fizz" (type/generate-string t (model/create-fizz-buzz-value 3))))
      (is (= "Buzz" (type/generate-string t (model/create-fizz-buzz-value 5))))
      (is (= "FizzBuzz" (type/generate-string t (model/create-fizz-buzz-value 15)))))))

(deftest fizz-buzz-type02-test
  (testing "タイプ 2: 数字のみ"
    (let [t (type/->FizzBuzzType02)]
      (is (= "1" (type/generate-string t (model/create-fizz-buzz-value 1))))
      (is (= "3" (type/generate-string t (model/create-fizz-buzz-value 3))))
      (is (= "5" (type/generate-string t (model/create-fizz-buzz-value 5))))
      (is (= "15" (type/generate-string t (model/create-fizz-buzz-value 15)))))))

(deftest fizz-buzz-type03-test
  (testing "タイプ 3: FizzBuzz の場合のみ"
    (let [t (type/->FizzBuzzType03)]
      (is (= "" (type/generate-string t (model/create-fizz-buzz-value 1))))
      (is (= "Fizz" (type/generate-string t (model/create-fizz-buzz-value 3))))
      (is (= "Buzz" (type/generate-string t (model/create-fizz-buzz-value 5))))
      (is (= "FizzBuzz" (type/generate-string t (model/create-fizz-buzz-value 15)))))))

(deftest fizz-buzz-type-not-defined-test
  (testing "タイプ未定義: 例外を発生させる"
    (let [t (type/->FizzBuzzTypeNotDefined)]
      (is (thrown? clojure.lang.ExceptionInfo
                   (type/generate-string t (model/create-fizz-buzz-value 1)))))))

(deftest create-type-test
  (testing "ファクトリ関数でタイプを生成する"
    (is (instance? fizzbuzz.domain.type.FizzBuzzType01 (type/create-type 1)))
    (is (instance? fizzbuzz.domain.type.FizzBuzzType02 (type/create-type 2)))
    (is (instance? fizzbuzz.domain.type.FizzBuzzType03 (type/create-type 3)))
    (is (instance? fizzbuzz.domain.type.FizzBuzzTypeNotDefined (type/create-type 99)))))
